Terminal links no longer working for node processes

Type: <b>Bug</b>

WIP

VS Code version: Code 1.78.2 (b3e4e68a0bc097f0ae7907b217c1119af9e03435, 2023-05-10T14:47:05.613Z)
OS version: Linux x64 5.19.0-42-generic snap
Modes:
Sandboxed: Yes

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz (8 x 3346)|
|GPU Status|2d_canvas: unavailable_software<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: disabled_software<br>multiple_raster_threads: enabled_on<br>opengl: disabled_off<br>rasterization: disabled_software<br>raw_draw: disabled_off_ok<br>video_decode: disabled_software<br>video_encode: disabled_software<br>vulkan: disabled_off<br>webgl: unavailable_software<br>webgl2: unavailable_software<br>webgpu: disabled_off|
|Load (avg)|2, 2, 2|
|Memory (System)|31.07GB (24.11GB free)|
|Process Argv|--no-sandbox --force-user-env --unity-launch --crash-reporter-id 23d56de3-001d-431f-81e8-0c352e7ea82f|
|Screen Reader|no|
|VM|0%|
|DESKTOP_SESSION|ubuntu|
|XDG_CURRENT_DESKTOP|Unity|
|XDG_SESSION_DESKTOP|ubuntu|
|XDG_SESSION_TYPE|x11|
</details><details><summary>Extensions (70)</summary>

Extension|Author (truncated)|Version
---|---|---
toggle-excluded-files|amo|2.0.0
tsl-problem-matcher|amo|0.6.2
vscode-zipfs|arc|3.0.0
chronicler|arc|0.1.16
better-toml|bun|0.3.2
css-theme-completions|con|0.0.5
esbuild-problem-matchers|con|0.0.2
vscode-eslint|dba|2.4.0
gitlens|eam|13.6.0
EditorConfig|Edi|0.16.4
RunOnSave|eme|0.2.0
json-tools|eri|1.0.2
prettier-vscode|esb|9.13.0
copilot|Git|1.87.127
copilot-labs|Git|0.14.884
vscode-github-actions|git|0.25.7
vscode-pull-request-github|Git|0.65.2023053009
vscode-mocha-test-adapter|hbe|2.14.1
vscode-test-explorer|hbe|2.21.1
rest-client|hum|0.25.1
template-string-converter|meg|0.6.1
git-graph|mhu|1.30.0
compare-folders|mos|0.22.5
vscode-json5|mrm|1.0.0
vscode-docker|ms-|1.25.1
vscode-language-pack-de|MS-|1.78.2023051009
python|ms-|2023.8.0
vscode-pylance|ms-|2023.5.40
jupyter|ms-|2023.4.1011241018
jupyter-keymap|ms-|1.1.0
jupyter-renderers|ms-|1.0.15
vscode-jupyter-cell-tags|ms-|0.1.8
vscode-jupyter-slideshow|ms-|0.1.5
remote-containers|ms-|0.292.0
remote-ssh|ms-|0.102.0
remote-ssh-edit|ms-|0.86.0
remote-wsl|ms-|0.78.8
vscode-remote-extensionpack|ms-|0.24.0
azure-account|ms-|0.11.5
cmake-tools|ms-|1.14.31
cpptools|ms-|1.15.4
cpptools-extension-pack|ms-|1.3.0
hexeditor|ms-|1.9.11
live-server|ms-|0.4.8
makefile-tools|ms-|0.7.0
powershell|ms-|2023.5.0
remote-explorer|ms-|0.4.0
remote-server|ms-|1.1.0
test-adapter-converter|ms-|0.1.7
vscode-js-profile-flame|ms-|1.0.5
hide-gitignored|npx|1.1.0
vetur|oct|0.37.3
vscode-twoslash-queries|Ort|1.1.0
typescript-mono-repo-import-helper|q|0.0.6
tsserver-live-reload|rbu|1.0.1
java|red|1.18.0
vscode-sort-json|ric|1.20.0
rust-analyzer|rus|0.3.1533
gitconfig|sid|2.0.1
cmake|twx|0.0.17
sort-lines|Tyr|1.10.1
intellicode-api-usage-examples|Vis|0.2.7
vscodeintellicode|Vis|1.2.30
vscode-java-dependency|vsc|0.22.0
vscode-java-pack|vsc|0.25.11
vscode-maven|vsc|0.41.0
vscode-icons|vsc|12.4.0
vscode-todo-highlight|way|1.0.5
vitest-explorer|Zix|0.2.41
config-editor|zwa|0.0.12

(1 theme extensions excluded)

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
python383cf:30185419
vspor879:30202332
vspor708:30202333
vspor363:30204092
vswsl492:30256859
vslsvsres303:30308271
vserr242cf:30382550
pythontb:30283811
vsjup518:30340749
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
vscorecescf:30445987
vscod805cf:30301675
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
cmake_vspar411:30581797
vsaa593:30376534
pythonvs932:30410667
cppdebug:30492333
vsclangdc:30486549
c4g48928:30535728
dsvsc012:30540252
pynewext54:30695312
azure-dev_surveyone:30548225
3biah626:30602489
pyind779:30671433
89544117:30613380
pythonsymbol12:30671437
a9j8j154:30646983
showlangstatbar:30737416
vsctsb:30748421
pythonms35:30701012
pythonfmttext:30731395
pythoncmv:30739700
fixhidewlkth:30730051
pythongtdpathcf:30739705

```

</details>

<!-- generated by issue reporter -->

Links are detected it seems:
```
2023-06-27 10:48:07.764 [trace] terminalMultiLineLinkDetector#detect text ["src/lib/driver/Driver.ts:324:36 - error TS18050: The value 'null' cannot be used here."]
2023-06-27 10:48:07.764 [trace] terminalMultiLineLinkDetector#detect candidate ["324:36"]
2023-06-27 10:48:07.764 [trace] terminalLocalLinkDetector#detect text ["src/lib/driver/Driver.ts:324:36 - error TS18050: The value 'null' cannot be used here."]
2023-06-27 10:48:07.765 [trace] terminalLocalLinkDetector#detect parsedLinks [[{"path":{"index":0,"text":"src/lib/driver/Driver.ts"},"suffix":{"row":324,"col":36,"suffix":{"index":24,"text":":324:36"}}}]]
2023-06-27 10:48:07.765 [trace] terminalLinkHelpers#updateLinkWithRelativeCwd cwd [null]
2023-06-27 10:48:07.765 [trace] terminalLocalLinkDetector#detect linkCandidates [["src/lib/driver/Driver.ts"]]
2023-06-27 10:48:07.765 [trace] terminalMultiLineLinkDetector#detect text ["src/lib/driver/Driver.ts:324:36 - error TS18050: The value 'null' cannot be used here."]
2023-06-27 10:48:07.766 [trace] terminalMultiLineLinkDetector#detect candidate ["324:36"]
2023-06-27 10:48:07.766 [trace] terminalLocalLinkDetector#detect text ["src/lib/driver/Driver.ts:324:36 - error TS18050: The value 'null' cannot be used here."]
2023-06-27 10:48:07.766 [trace] terminalLocalLinkDetector#detect parsedLinks [[{"path":{"index":0,"text":"src/lib/driver/Driver.ts"},"suffix":{"row":324,"col":36,"suffix":{"index":24,"text":":324:36"}}}]]
2023-06-27 10:48:07.766 [trace] terminalLinkHelpers#updateLinkWithRelativeCwd cwd [null]
2023-06-27 10:48:07.766 [trace] terminalLocalLinkDetector#detect linkCandidates [["src/lib/driver/Driver.ts"]]
2023-06-27 10:48:08.680 [trace] terminalMultiLineLinkDetector#detect text ["src/lib/driver/Driver.ts:324:36 - error TS18050: The value 'null' cannot be used here."]
2023-06-27 10:48:08.681 [trace] terminalMultiLineLinkDetector#detect candidate ["324:36"]
2023-06-27 10:48:08.681 [trace] terminalLocalLinkDetector#detect text ["src/lib/driver/Driver.ts:324:36 - error TS18050: The value 'null' cannot be used here."]
2023-06-27 10:48:08.681 [trace] terminalLocalLinkDetector#detect parsedLinks [[{"path":{"index":0,"text":"src/lib/driver/Driver.ts"},"suffix":{"row":324,"col":36,"suffix":{"index":24,"text":":324:36"}}}]]
2023-06-27 10:48:08.681 [trace] terminalLinkHelpers#updateLinkWithRelativeCwd cwd [null]
2023-06-27 10:48:08.682 [trace] terminalLocalLinkDetector#detect linkCandidates [["src/lib/driver/Driver.ts"]]
2023-06-27 10:48:08.682 [trace] terminalMultiLineLinkDetector#detect text ["src/lib/driver/Driver.ts:324:36 - error TS18050: The value 'null' cannot be used here."]
2023-06-27 10:48:08.682 [trace] terminalMultiLineLinkDetector#detect candidate ["324:36"]
2023-06-27 10:48:08.682 [trace] terminalLocalLinkDetector#detect text ["src/lib/driver/Driver.ts:324:36 - error TS18050: The value 'null' cannot be used here."]
2023-06-27 10:48:08.683 [trace] terminalLocalLinkDetector#detect parsedLinks [[{"path":{"index":0,"text":"src/lib/driver/Driver.ts"},"suffix":{"row":324,"col":36,"suffix":{"index":24,"text":":324:36"}}}]]
2023-06-27 10:48:08.683 [trace] terminalLinkHelpers#updateLinkWithRelativeCwd cwd [null]
2023-06-27 10:48:08.683 [trace] terminalLocalLinkDetector#detect linkCandidates [["src/lib/driver/Driver.ts"]]
2023-06-27 10:48:08.865 [trace] terminalMultiLineLinkDetector#detect text ["[10:47:33 AM] Starting compilation in watch mode..."]
2023-06-27 10:48:08.865 [trace] terminalMultiLineLinkDetector#detect candidate ["10:47"]
2023-06-27 10:48:08.865 [trace] terminalLocalLinkDetector#detect text ["[10:47:33 AM] Starting compilation in watch mode..."]
2023-06-27 10:48:08.866 [trace] terminalLocalLinkDetector#detect parsedLinks [[{"path":{"index":1,"text":"10"},"suffix":{"row":47,"col":33,"suffix":{"index":3,"text":":47:33"}}}]]
2023-06-27 10:48:08.866 [trace] terminalLinkHelpers#updateLinkWithRelativeCwd cwd [null]
2023-06-27 10:48:08.866 [trace] terminalLocalLinkDetector#detect linkCandidates [["10"]]
2023-06-27 10:48:08.866 [trace] terminalMultiLineLinkDetector#detect text ["[10:47:33 AM] Starting compilation in watch mode..."]
2023-06-27 10:48:08.867 [trace] terminalMultiLineLinkDetector#detect candidate ["10:47"]
2023-06-27 10:48:08.867 [trace] terminalLocalLinkDetector#detect text ["[10:47:33 AM] Starting compilation in watch mode..."]
2023-06-27 10:48:08.867 [trace] terminalLocalLinkDetector#detect parsedLinks [[{"path":{"index":1,"text":"10"},"suffix":{"row":47,"col":33,"suffix":{"index":3,"text":":47:33"}}}]]
2023-06-27 10:48:08.867 [trace] terminalLinkHelpers#updateLinkWithRelativeCwd cwd [null]
2023-06-27 10:48:08.867 [trace] terminalLocalLinkDetector#detect linkCandidates [["10"]]
2023-06-27 10:48:16.002 [trace] terminalMultiLineLinkDetector#detect text ["src/lib/driver/Driver.ts:324:36 - error TS18050: The value 'null' cannot be used here."]
2023-06-27 10:48:16.002 [trace] terminalMultiLineLinkDetector#detect candidate ["324:36"]
2023-06-27 10:48:16.002 [trace] terminalLocalLinkDetector#detect text ["src/lib/driver/Driver.ts:324:36 - error TS18050: The value 'null' cannot be used here."]
2023-06-27 10:48:16.002 [trace] terminalLocalLinkDetector#detect parsedLinks [[{"path":{"index":0,"text":"src/lib/driver/Driver.ts"},"suffix":{"row":324,"col":36,"suffix":{"index":24,"text":":324:36"}}}]]
2023-06-27 10:48:16.002 [trace] terminalLinkHelpers#updateLinkWithRelativeCwd cwd [null]
2023-06-27 10:48:16.002 [trace] terminalLocalLinkDetector#detect linkCandidates [["src/lib/driver/Driver.ts"]]
2023-06-27 10:48:16.003 [trace] terminalMultiLineLinkDetector#detect text ["src/lib/driver/Driver.ts:324:36 - error TS18050: The value 'null' cannot be used here."]
2023-06-27 10:48:16.003 [trace] terminalMultiLineLinkDetector#detect candidate ["324:36"]
2023-06-27 10:48:16.003 [trace] terminalLocalLinkDetector#detect text ["src/lib/driver/Driver.ts:324:36 - error TS18050: The value 'null' cannot be used here."]
2023-06-27 10:48:16.003 [trace] terminalLocalLinkDetector#detect parsedLinks [[{"path":{"index":0,"text":"src/lib/driver/Driver.ts"},"suffix":{"row":324,"col":36,"suffix":{"index":24,"text":":324:36"}}}]]
2023-06-27 10:48:16.003 [trace] terminalLinkHelpers#updateLinkWithRelativeCwd cwd [null]
2023-06-27 10:48:16.003 [trace] terminalLocalLinkDetector#detect linkCandidates [["src/lib/driver/Driver.ts"]]
2023-06-27 10:48:16.235 [trace] terminalMultiLineLinkDetector#detect text ["src/lib/driver/Driver.ts:324:36 - error TS18050: The value 'null' cannot be used here."]
2023-06-27 10:48:16.235 [trace] terminalMultiLineLinkDetector#detect candidate ["324:36"]
2023-06-27 10:48:16.236 [trace] terminalLocalLinkDetector#detect text ["src/lib/driver/Driver.ts:324:36 - error TS18050: The value 'null' cannot be used here."]
2023-06-27 10:48:16.236 [trace] terminalLocalLinkDetector#detect parsedLinks [[{"path":{"index":0,"text":"src/lib/driver/Driver.ts"},"suffix":{"row":324,"col":36,"suffix":{"index":24,"text":":324:36"}}}]]
2023-06-27 10:48:16.236 [trace] terminalLinkHelpers#updateLinkWithRelativeCwd cwd [null]
2023-06-27 10:48:16.236 [trace] terminalLocalLinkDetector#detect linkCandidates [["src/lib/driver/Driver.ts"]]
2023-06-27 10:48:16.236 [trace] terminalMultiLineLinkDetector#detect text ["src/lib/driver/Driver.ts:324:36 - error TS18050: The value 'null' cannot be used here."]
2023-06-27 10:48:16.237 [trace] terminalMultiLineLinkDetector#detect candidate ["324:36"]
2023-06-27 10:48:16.238 [trace] terminalLocalLinkDetector#detect text ["src/lib/driver/Driver.ts:324:36 - error TS18050: The value 'null' cannot be used here."]
2023-06-27 10:48:16.238 [trace] terminalLocalLinkDetector#detect parsedLinks [[{"path":{"index":0,"text":"src/lib/driver/Driver.ts"},"suffix":{"row":324,"col":36,"suffix":{"index":24,"text":":324:36"}}}]]
2023-06-27 10:48:16.238 [trace] terminalLinkHelpers#updateLinkWithRelativeCwd cwd [null]
2023-06-27 10:48:16.238 [trace] terminalLocalLinkDetector#detect linkCandidates [["src/lib/driver/Driver.ts"]]
2023-06-27 10:48:18.285 [trace] terminalMultiLineLinkDetector#detect text ["[10:47:33 AM] Starting compilation in watch mode..."]
2023-06-27 10:48:18.286 [trace] terminalMultiLineLinkDetector#detect candidate ["10:47"]
2023-06-27 10:48:18.286 [trace] terminalLocalLinkDetector#detect text ["[10:47:33 AM] Starting compilation in watch mode..."]
2023-06-27 10:48:18.286 [trace] terminalLocalLinkDetector#detect parsedLinks [[{"path":{"index":1,"text":"10"},"suffix":{"row":47,"col":33,"suffix":{"index":3,"text":":47:33"}}}]]
2023-06-27 10:48:18.286 [trace] terminalLinkHelpers#updateLinkWithRelativeCwd cwd [null]
2023-06-27 10:48:18.287 [trace] terminalLocalLinkDetector#detect linkCandidates [["10"]]
2023-06-27 10:48:18.287 [trace] terminalMultiLineLinkDetector#detect text ["[10:47:33 AM] Starting compilation in watch mode..."]
2023-06-27 10:48:18.287 [trace] terminalMultiLineLinkDetector#detect candidate ["10:47"]
2023-06-27 10:48:18.287 [trace] terminalLocalLinkDetector#detect text ["[10:47:33 AM] Starting compilation in watch mode..."]
2023-06-27 10:48:18.288 [trace] terminalLocalLinkDetector#detect parsedLinks [[{"path":{"index":1,"text":"10"},"suffix":{"row":47,"col":33,"suffix":{"index":3,"text":":47:33"}}}]]
2023-06-27 10:48:18.288 [trace] terminalLinkHelpers#updateLinkWithRelativeCwd cwd [null]
2023-06-27 10:48:18.288 [trace] terminalLocalLinkDetector#detect linkCandidates [["10"]]
```

But I cannot CTRL+click them - nothing happens, no underline, no navigation.

--- a/src/vs/workbench/api/browser/mainThreadTerminalService.ts
+++ b/src/vs/workbench/api/browser/mainThreadTerminalService.ts
@@ -43,13 +43,14 @@ export class MainThreadTerminalService implements MainThreadTerminalServiceShape
 	private readonly _quickFixProviders = new Map<string, IDisposable>();
 	private readonly _dataEventTracker = new MutableDisposable<TerminalDataEventTracker>();
 	private readonly _sendCommandEventListener = new MutableDisposable();
+
 	/**
 	 * A single shared terminal link provider for the exthost. When an ext registers a link
 	 * provider, this is registered with the terminal on the renderer side and all links are
 	 * provided through this, even from multiple ext link providers. Xterm should remove lower
 	 * priority intersecting links itself.
 	 */
-	private _linkProvider: IDisposable | undefined;
+	private _linkProvider = this._store.add(new MutableDisposable());
 
 	private _os: OperatingSystem = OS;
 
@@ -111,7 +112,6 @@ export class MainThreadTerminalService implements MainThreadTerminalServiceShape
 
 	public dispose(): void {
 		this._store.dispose();
-		this._linkProvider?.dispose();
 		for (const provider of this._profileProviders.values()) {
 			provider.dispose();
 		}
@@ -253,13 +253,11 @@ export class MainThreadTerminalService implements MainThreadTerminalServiceShape
 	}
 
 	public $startLinkProvider(): void {
-		this._linkProvider?.dispose();
-		this._linkProvider = this._terminalLinkProviderService.registerLinkProvider(new ExtensionTerminalLinkProvider(this._proxy));
+		this._linkProvider.value = this._terminalLinkProviderService.registerLinkProvider(new ExtensionTerminalLinkProvider(this._proxy));
 	}
 
 	public $stopLinkProvider(): void {
-		this._linkProvider?.dispose();
-		this._linkProvider = undefined;
+		this._linkProvider.clear();
 	}
 
 	public $registerProcessSupport(isSupported: boolean): void {

--- a/src/vs/workbench/contrib/terminalContrib/links/browser/terminal.links.contribution.ts
+++ b/src/vs/workbench/contrib/terminalContrib/links/browser/terminal.links.contribution.ts
@@ -66,15 +66,15 @@ class TerminalLinkContribution extends DisposableStore implements ITerminalContr
 		// Attach the link provider(s) to the instance and listen for changes
 		if (!isDetachedTerminalInstance(this._instance)) {
 			for (const linkProvider of this._terminalLinkProviderService.linkProviders) {
-				this._linkManager.registerExternalLinkProvider(linkProvider.provideLinks.bind(linkProvider, this._instance));
+				linkManager.registerExternalLinkProvider(linkProvider.provideLinks.bind(linkProvider, this._instance));
 			}
-			this._linkManager.add(this._terminalLinkProviderService.onDidAddLinkProvider(e => {
+			linkManager.add(this._terminalLinkProviderService.onDidAddLinkProvider(e => {
 				linkManager.registerExternalLinkProvider(e.provideLinks.bind(e, this._instance as ITerminalInstance));
 			}));
 		}
 
 		// TODO: Currently only a single link provider is supported; the one registered by the ext host
-		this._linkManager.add(this._terminalLinkProviderService.onDidRemoveLinkProvider(e => {
+		linkManager.add(this._terminalLinkProviderService.onDidRemoveLinkProvider(e => {
 			linkManager.dispose();
 			this.xtermReady(xterm);
 		}));

--- a/src/vs/workbench/contrib/terminalContrib/links/browser/terminal.links.contribution.ts
+++ b/src/vs/workbench/contrib/terminalContrib/links/browser/terminal.links.contribution.ts
@@ -3,6 +3,7 @@
  *  Licensed under the MIT License. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
+import { Event } from 'vs/base/common/event';
 import { KeyCode, KeyMod } from 'vs/base/common/keyCodes';
 import { DisposableStore } from 'vs/base/common/lifecycle';
 import { localize2 } from 'vs/nls';
@@ -50,27 +51,30 @@ class TerminalLinkContribution extends DisposableStore implements ITerminalContr
 	}
 
 	xtermReady(xterm: IXtermTerminal & { raw: RawXtermTerminal }): void {
-		const linkManager = this._instantiationService.createInstance(TerminalLinkManager, xterm.raw, this._processManager, this._instance.capabilities, this._linkResolver);
+		const linkManager = this._linkManager = this.add(this._instantiationService.createInstance(TerminalLinkManager, xterm.raw, this._processManager, this._instance.capabilities, this._linkResolver));
+
+		// Set widget manager
 		if (isTerminalProcessManager(this._processManager)) {
-			this._processManager.onProcessReady(() => {
+			const disposable = this.add(Event.once(this._processManager.onProcessReady)(() => {
 				linkManager.setWidgetManager(this._widgetManager);
-			});
+				this.delete(disposable);
+			}));
 		} else {
 			linkManager.setWidgetManager(this._widgetManager);
 		}
-		this._linkManager = this.add(linkManager);
 
 		// Attach the link provider(s) to the instance and listen for changes
 		if (!isDetachedTerminalInstance(this._instance)) {
 			for (const linkProvider of this._terminalLinkProviderService.linkProviders) {
 				this._linkManager.registerExternalLinkProvider(linkProvider.provideLinks.bind(linkProvider, this._instance));
 			}
-			this.add(this._terminalLinkProviderService.onDidAddLinkProvider(e => {
+			this._linkManager.add(this._terminalLinkProviderService.onDidAddLinkProvider(e => {
 				linkManager.registerExternalLinkProvider(e.provideLinks.bind(e, this._instance as ITerminalInstance));
 			}));
 		}
+
 		// TODO: Currently only a single link provider is supported; the one registered by the ext host
-		this.add(this._terminalLinkProviderService.onDidRemoveLinkProvider(e => {
+		this._linkManager.add(this._terminalLinkProviderService.onDidRemoveLinkProvider(e => {
 			linkManager.dispose();
 			this.xtermReady(xterm);
 		}));

--- a/src/vs/workbench/contrib/terminalContrib/links/browser/terminalLinkDetectorAdapter.ts
+++ b/src/vs/workbench/contrib/terminalContrib/links/browser/terminalLinkDetectorAdapter.ts
@@ -76,7 +76,7 @@ export class TerminalLinkDetectorAdapter extends Disposable implements ILinkProv
 		// Cap the maximum context on either side of the line being provided, by taking the context
 		// around the line being provided for this ensures the line the pointer is on will have
 		// links provided.
-		const maxLineContext = Math.max(this._detector.maxLinkLength / this._detector.xterm.cols);
+		const maxLineContext = Math.max(this._detector.maxLinkLength, this._detector.xterm.cols);
 		const minStartLine = Math.max(startLine - maxLineContext, 0);
 		const maxEndLine = Math.min(endLine + maxLineContext, this._detector.xterm.buffer.active.length);
 

--- a/src/vs/workbench/contrib/terminalContrib/links/browser/terminalLinkManager.ts
+++ b/src/vs/workbench/contrib/terminalContrib/links/browser/terminalLinkManager.ts
@@ -91,6 +91,8 @@ export class TerminalLinkManager extends DisposableStore {
 		let activeHoverDisposable: IDisposable | undefined;
 		let activeTooltipScheduler: RunOnceScheduler | undefined;
 		this.add(toDisposable(() => {
+			this._clearLinkProviders();
+			dispose(this._externalLinkProviders);
 			activeHoverDisposable?.dispose();
 			activeTooltipScheduler?.dispose();
 		}));

--- a/src/vs/workbench/contrib/terminalContrib/links/browser/terminalLinkManager.ts
+++ b/src/vs/workbench/contrib/terminalContrib/links/browser/terminalLinkManager.ts
@@ -352,7 +352,11 @@ export class TerminalLinkManager extends DisposableStore {
 		}
 	}
 
-	registerExternalLinkProvider(provideLinks: OmitFirstArg<ITerminalExternalLinkProvider['provideLinks']>): IDisposable {
+	registerExternalLinkProvider(provideLinks: OmitFirstArg<ITerminalExternalLinkProvider['provideLinks']>): void {
+		// Avoid any leaks in case this is already disposed
+		if (this.isDisposed) {
+			return;
+		}
 		// Clear and re-register the standard link providers so they are a lower priority than the new one
 		this._clearLinkProviders();
 		const detectorId = `extension-${this._externalLinkProviders.length}`;
@@ -360,7 +364,6 @@ export class TerminalLinkManager extends DisposableStore {
 		const newLinkProvider = this._xterm.registerLinkProvider(wrappedLinkProvider);
 		this._externalLinkProviders.push(newLinkProvider);
 		this._registerStandardLinkProviders();
-		return newLinkProvider;
 	}
 
 	protected _isLinkActivationModifierDown(event: MouseEvent): boolean {

--- a/src/vs/workbench/contrib/terminalContrib/links/browser/terminalLinkResolver.ts
+++ b/src/vs/workbench/contrib/terminalContrib/links/browser/terminalLinkResolver.ts
@@ -15,8 +15,6 @@ import { ITerminalBackend } from 'vs/platform/terminal/common/terminal';
 import { mainWindow } from 'vs/base/browser/window';
 
 export class TerminalLinkResolver implements ITerminalLinkResolver {
-	declare _serviceBrand: undefined;
-
 	// Link cache could be shared across all terminals, but that could lead to weird results when
 	// both local and remote terminals are present
 	private readonly _resolvedLinkCaches: Map<string, LinkCache> = new Map();

--- a/src/vs/workbench/contrib/terminalContrib/links/test/browser/terminalLinkManager.test.ts
+++ b/src/vs/workbench/contrib/terminalContrib/links/test/browser/terminalLinkManager.test.ts
@@ -99,6 +99,14 @@ suite('TerminalLinkManager', () => {
 		} as Partial<ITerminalCapabilityStore> as any, instantiationService.createInstance(TerminalLinkResolver)));
 	});
 
+	suite('registerExternalLinkProvider', () => {
+		test('should not leak disposables if the link manager is already disposed', () => {
+			linkManager.registerExternalLinkProvider(async () => undefined);
+			linkManager.dispose();
+			linkManager.registerExternalLinkProvider(async () => undefined);
+		});
+	});
+
 	suite('getLinks and open recent link', () => {
 		test('should return no links', async () => {
 			const links = await linkManager.getLinks();

